src/Symfony/Component/Routing/Generator/CompiledUrlGenerator.php

```diff
@@ -53,6 +52,16 @@ public function generate($name, $parameters = [], $referenceType = self::ABSOLUT
 
         list($variables, $defaults, $requirements, $tokens, $hostTokens, $requiredSchemes) = $this->compiledRoutes[$name];
 
+        if (isset($defaults['_canonical_route']) && isset($defaults['_locale'])) {
+            if (\in_array('_locale', $variables, true)) {
+                if (!isset($parameters['_locale'])) {
+                    $parameters['_locale'] = $defaults['_locale'];
+                }
+            } else {
+                unset($parameters['_locale']);
```

Comment:
We always unset the _locale parameter if the url does not have the _locale variable because the good i18n route has already been found and we don't want the parameter.